### DISCUSSION TITLE
Can't get the guilds of the current user (Discord)

### DISCUSSION THREAD

**Discussion by griffan113** [nextauthjs/next-auth#1629] [Help] [answered]
I'm trying to get the guilds that the user is on

 Hello, i'm making my bot's dashboard, buy and i got the following error: 

(node:60208) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'accessToken' of null

The database and the auth system are working, but I can't proceed because of this error.

![image](https://user-images.githubusercontent.com/70238970/113207233-c119ed80-9246-11eb-96a6-2cec5597a644.png)

Source Code Screenshots:
getuserguilds.ts:

![image](https://user-images.githubusercontent.com/70238970/113207378-ec044180-9246-11eb-9e46-326fce3abaa6.png)

[...nextauth].js

![image](https://user-images.githubusercontent.com/70238970/113207561-21a92a80-9247-11eb-9849-f87482cf013b.png)

* [ ] Found the documentation helpful
* [X] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by nodgear** [comment, marked as answer]
@ Edit: the solution is under this comment.
The discord provider only gives the default information mentioned on the next-auth api, but everything is present on the callbacks, solving a promise while injecting the data on token callback allows us to get it from the session context

**Reply by griffan113** [reply]
Yes, unfortunately this documentation isn't complete yet, I had to make a manual back-end with Node.js.

**Reply by nodgear** [reply]
Well, here's the solution. Good luck with your project!
```javascript
 // Pass session jwt: true
  session: {
    jwt: true,
  },
// Also pass these callbacks
  callbacks: {
    jwt: async (token, user, account, profile) => {
      // Add auth_time to token on signin in
      if (user) {
        token.accessToken = account.accessToken;
        token.refreshToken = account.refreshToken;
        token.id = profile.id;
        token.discriminator = profile.discriminator;
      }
      return Promise.resolve(token);
    },
    session: async (session, token) => {
      session.user.accessToken = token.accessToken;
      session.user.refreshToken = token.refreshToken;
      session.user.id = token.id;
      session.user.discriminator = token.discriminator;

      return session;
    }
    }
```

**Reply by griffan113** [reply]
Many thx man :)

**Reply by ariscript** [reply]
This is really useful, the docs for getting extra information really does need a good rework. Also @griffan113, you can just fetch the guilds in the `session` callback and put them in the session object, which is way more convenient to use than having to fetch an API route *every single time*. This is how i implemented these callbacks:
```ts
jwt: async (token, user, account, profile) => {
    // Add auth_time to token on signin in
    if (user) {
        token.profile = profile;
        token.accessToken = account.accessToken;
        token.refreshToken = account.refreshToken;
    }
    return Promise.resolve(token);
},
// @ts-ignore
session: async (session, token) => {
    // @ts-ignore
    session.user.accessToken = token.accessToken;
    // @ts-ignore
    session.user.refreshToken = token.refreshToken;
    // @ts-ignore
    session.user.profile = token.profile;

    const res = await fetch(
        "https://discord.com/api/v8/users/@me/guilds",
        {
            // @ts-ignore
            headers: { Authorization: `Bearer ${token.accessToken}` },
        }
    );
    const data = await res.json();
    // @ts-ignore
    session.user.profile.guilds = data;

    return session;
},
```
If you don't use typescript, you can safely ignore the `ts-ignore` comments.

**Reply by griffan113** [reply]
Wow, many thx, man I'll study this way, and yes I am using TS.

**Reply by balazsorban44** [reply]
Please don't fetch data in the session callback, as it will be an expensive operation, running every time you ask for the session from the session endpoint!

Check out the profile callback on the provider, or use the if block of your jwt callback!

https://next-auth.js.org/configuration/providers#using-a-custom-provider

**Reply by nodgear** [reply]
I can't see how fetching the data is a better implementation than just grabbing the values from the JWT callback.

But i have to agree the documentation for the discord provider could get some love.

**Reply by balazsorban44** [reply]
If I understand you correctly, yes fetch is only needed if you could not receive something from your provider  in the first request. If you have, just use it and return it in `jwt`, obviously. I think here it was done BECAUSE something was needed that was missing in the provider response.

If you think the docs could be improved, feel free to open a PR with your suggestions anytime! We welcome contributions 😊
